esinstall/src/entrypoints.ts

```diff
@@ -0,0 +1,337 @@
+    const exportField = (packageManifest as PackageManifestWithExports).exports;
+
+    // If this is a non-main entry point
+    if (packageEntrypoint) {
+      const normalizedMap = explodeExportMap(exportField, {
+        cwd: path.dirname(packageManifestLoc),
+      });
+      const mapValue = normalizedMap && Reflect.get(normalizedMap, './' + packageEntrypoint);
+
+      if (typeof mapValue !== 'string') {
+        throw new Error(
+          `Package "${packageName}" exists but package.json "exports" does not include entry for "./${packageEntrypoint}".`,
+        );
+      }
+      return path.join(packageManifestLoc, '..', mapValue);
+    } else {
+      const exportMapEntry = exportField['.'];
+      const mapValue = findExportMapEntry(exportMapEntry);
+
+      if (mapValue) {
+        return path.join(packageManifestLoc, '..', mapValue);
+      }
+    }
+  }
+
+  // if, no export map and dep points directly to a file within a package, return that reference.
+  if (path.extname(dep) && !validatePackageName(dep).validForNewPackages) {
+    return realpathSync.native(require.resolve(dep, {paths: [cwd]}));
+  }
+
+  // Otherwise, resolve directly to the dep specifier. Note that this supports both
+  // "package-name" & "package-name/some/path" where "package-name/some/path/package.json"
+  // exists at that lower path, that must be used to resolve. In that case, export
+  // maps should not be supported.
+  const [depManifestLoc, depManifest] = resolveDependencyManifest(dep, cwd);
+
+  if (!depManifest) {
+    try {
+      const maybeLoc = realpathSync.native(require.resolve(dep, {paths: [cwd]}));
+      return maybeLoc;
+    } catch {
+      // Oh well, was worth a try
+    }
+  }
+
+  if (!depManifestLoc || !depManifest) {
+    throw new Error(
+      `Package "${dep}" not found. Have you installed it? ${depManifestLoc ? depManifestLoc : ''}`,
+    );
+  }
+
+  let foundEntrypoint = findManifestEntry(depManifest, dep, {
+    packageName,
+    packageLookupFields,
+  });
+
+  // Some packages are types-only. If this is one of those packages, resolve with that.
+  if (!foundEntrypoint && hasTypes(depManifest)) {
+    const typesLoc = (depManifest.types || depManifest.typings) as string;
+    return path.join(depManifestLoc, '..', typesLoc);
+  }
+  // Sometimes packages don't give an entrypoint, assuming you'll fall back to "index.js".
+  if (!foundEntrypoint) {
+    foundEntrypoint = 'index.js';
+  }
+  if (typeof foundEntrypoint !== 'string') {
+    throw new Error(`"${dep}" has unexpected entrypoint: ${JSON.stringify(foundEntrypoint)}.`);
+  }
+
+  return realpathSync.native(
+    require.resolve(path.join(depManifestLoc || '', '..', foundEntrypoint)),
+  );
+}
+
+const picoMatchGlobalOptions = Object.freeze({
+  capture: true,
+  noglobstar: true,
+});
+
+function* forEachExportEntry(
+  exportField: ExportField,
+): Generator<[string, unknown], any, undefined> {
```

Comment:
A generator! ✨

Reply:
oh no does this mean I need to learn how Generators work???

Reply:
Happy to explain to anyone, I find them great for simplifying apis where complex logic can be reduced to a loop.